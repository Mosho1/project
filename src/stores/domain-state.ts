import {
    types,
    getEnv,
    applySnapshot,
    getSnapshot,
    detach,
    hasParent,
    getParent
} from 'mobx-state-tree'
import { pouch } from './utils/pouchdb-model';
import { models, modelTypes } from './models/index';
import { ContextMenu } from './context-menu';
import { ICodeBlock, ICodeBlockIO } from './models/code-block';
import * as codeBlocks from './functions';
import { values } from './utils/utils';
import { run, stop } from './run';
import { SocketTypeEnum, areSocketsCompatible } from './models/socket';
import { observable } from 'mobx';

const Position = types.model('Position', {
    x: types.number,
    y: types.number
});

const Stage = types.model('Stage', {
    scale: types.optional(types.number, 1),
    position: types.optional(Position, { x: 0, y: 0 })
}).views(self => ({
    get store(): null | IStore {
        if (!hasParent(self, 1)) return null;
        return getParent(self, 1);
    },
})).actions(self => ({
    move(dx: number, dy: number) {
        self.position.x += dx;
        self.position.y += dy;
        return self;
    },
    setScale(scale: number) {
        self.scale = scale;
    },
    handleScale(e: WheelEvent) {
        const mousePointTo = {
            x: e.clientX / self.scale - self.position.x / self.scale,
            y: e.clientY / self.scale - self.position.y / self.scale,
        };

        const scaleBy = e.ctrlKey ? 1.3 : 1.1;

        const newScale = e.deltaY < 0 ? self.scale * scaleBy : self.scale / scaleBy;

        if (newScale > 3 || newScale < 0.4) return;

        self.scale = newScale;

        self.position = {
            x: -(mousePointTo.x - e.clientX / newScale) * newScale,
            y: -(mousePointTo.y - e.clientY / newScale) * newScale
        }
    },
    getAbsolutePosition(x: number, y: number) {
        return {
            x: x - self.position.x,
            y: y - self.position.y,
        }
    }
}));

type IStageType = typeof Stage.Type;
export interface IStage extends IStageType { };

export const Store = pouch.store('Store', {
    boxes: types.optional(types.map(models.Box), {}),
    sockets: types.optional(types.map(models.Socket), {}),
    arrows: types.optional(types.map(models.Arrow), {}),
    codeBlocks: types.optional(types.map(models.CodeBlock), {}),
    selection: types.optional(types.array(types.reference<modelTypes['Box']>(models.Box)), []),
    draggedArrow: types.maybe(models.DraggedArrow),
    draggedRect: types.maybe(models.DraggedRect),
    draggedFromSocket: types.maybe(types.reference<modelTypes['Socket']>(models.Socket)),
    contextMenu: types.maybe(ContextMenu),
    stage: types.optional(Stage, {}),
}).volatile(_ => ({
    running: false,
    breakpointCallback: null as Function | null,
    breakPosition: null as modelTypes['Box'] | null
}))
    .actions(self => ({
        setDraggedFromSocket(socket: modelTypes['Socket'] | null) {
            self.draggedFromSocket = socket;
        },
        setBreakpoint(box: modelTypes['Box'], cb: Function) {
            self.breakpointCallback = cb;
            self.breakPosition = box;
        },
        continueAfterBreakpoint() {
            self.breakpointCallback = null;
            self.breakPosition = null;
        }

    }))
    .actions(self => {
        const addBox = (name: string, x: number, y: number, code: ICodeBlock) => {
            const { inputs, returns, execInputs, execOutputs, values } = code;
            const boxValues = values.map(v => ({ name: v.name, value: v.defaultValue || '' }));
            const box = models.Box.create({ name, code, values: boxValues, ...self.stage.getAbsolutePosition(x - 75, y - 60) });
            for (const input of inputs) {
                addSocketToBox(box, 'input', input);
            }
            if (returns && returns.type !== 'void') {
                addSocketToBox(box, 'output', returns);
            }
            for (const input of execInputs) {
                addSocketToBox(box, 'execInput', input);
            }
            for (const output of execOutputs) {
                addSocketToBox(box, 'execOutput', output);
            }
            self.boxes.put(box);
            return box;
        };
        const addBoxAndArrowIfDragged = (name: string, x: number, y: number, code: ICodeBlock) => {
            const b = addBox(name, x, y, code);
            if (self.draggedFromSocket) {
                for (const socket of b.sockets) {
                    if (areSocketsCompatible(socket, self.draggedFromSocket)) {
                        const a = addArrowFromDraggedSocket(socket);
                        if (a !== null) return true;
                    }
                }
            }
            return false;
        };
        const setSelection = (selection: modelTypes['Box'][]) => {
            self.selection.replace(selection);
            return self;
        };
        const addToSelection = (selection: modelTypes['Box'][]) => {
            self.selection.push(...selection);
            return self;
        };
        const deleteBox = (box: modelTypes['Box']) => {
            self.selection.remove(box);
            for (const socket of box.sockets) {
                deleteArrowsForSocket(socket);
            }

            for (const socket of box.sockets) {
                detach(socket);
            }

            detach(box);
        };
        const deleteSelection = () => {
            const boxes = self.selection.slice(0);
            for (const box of boxes) {
                deleteBox(box);
            }
            setSelection([]);
        };
        const createBox = (name: string, x: number, y: number, code: ICodeBlock) => {
            const box = addBox(name, x, y, code);
            // setSelection(box);
            return box;
        };
        const startDragArrow = (socket: modelTypes['Socket']) => {
            const { x, y } = socket;
            if (socket.isInput && socket.arrows.length > 0) return;
            self.draggedArrow = models.DraggedArrow.create({ startX: x, startY: y, endX: x, endY: y });
            self.setDraggedFromSocket(socket);
        };

        const moveDragArrow = (x: number, y: number) => {
            if (self.draggedArrow) {
                if (self.draggedFromSocket!.isInput) {
                    self.draggedArrow.start(x, y);
                } else {
                    self.draggedArrow.end(x, y);
                }
            }
        };
        const endDragArrow = (
            socket?: modelTypes['Socket'] | null,
            clientX = self.draggedArrow!.endX,
            clientY = self.draggedArrow!.endY,
        ) => {
            self.draggedArrow = null;
            if (!self.draggedFromSocket) return;
            if (socket) {
                addArrowFromDraggedSocket(socket);
            } else if (self.contextMenu) {
                const { isInput, isExec, code } = self.draggedFromSocket;
                let typeFilter: null | ((code: ICodeBlock) => boolean) = null;
                if (isExec) {
                    if (isInput) {
                        typeFilter = (c) => c.execOutputs.some(i => i.type === code.type);
                    } else {
                        typeFilter = (c) => c.execInputs.some(i => i.type === code.type);
                    }
                } else {
                    switch (code.type) {
                        case 'number':
                        case 'string':
                            if (isInput) {
                                typeFilter = (c) => c.returns ? c.returns.type === 'any' || c.returns.type === code.type : false;
                            } else {
                                typeFilter = (c) => c.inputs.some(i => i.type === 'any' || i.type === code.type);
                            }
                    }
                }

                setTimeout(() => self.contextMenu!.toggle(true, clientX, clientY, typeFilter));
            }
        };
        const startDragRect = (x: number, y: number) => {
            self.draggedRect = models.DraggedRect.create({ startX: x, startY: y, endX: x, endY: y });
        };

        const moveDragRect = (x: number, y: number) => {
            if (!self.draggedRect) return;
            self.draggedRect.end(x, y);
        };
        const endDragRect = () => {
            if (!self.draggedRect) return;
            const selection = observable<modelTypes['Box']>([]);
            for (const box of values(self.boxes)) {
                if (self.draggedRect.hasIntersection(box)) {
                    selection.push(box);
                }
            }
            self.selection = selection;
            self.draggedRect = null;
        };
        const hasArrow = (from: modelTypes['Socket'], to?: modelTypes['Socket']) => {
            const [input, output] = from.isInput ? [from, to] : [to, from];
            return Boolean(values(self.arrows).find(a => {
                return (!input || a.input === input) &&
                    (!output || a.output === output);
            }));
        };
        const addArrow = (input: modelTypes['Socket'], output: modelTypes['Socket']) => {
            if (input.isCompatibleWith(output) &&
                !hasArrow(input, output)) {
                let arrow: null | modelTypes['Arrow'] = null;
                arrow = models.Arrow.create({ input, output });
                self.arrows.put(arrow);
                return arrow;
            }

            return null;

        };
        const addArrowFromDraggedSocket = (socket: modelTypes['Socket']) => {
            if (!self.draggedFromSocket) return;
            const [input, output] = socket.isInput ? [socket, self.draggedFromSocket] : [self.draggedFromSocket, socket];
            self.setDraggedFromSocket(null);
            return addArrow(input, output);
        };
        const deleteArrowsForSocket = (socket: modelTypes['Socket']) => {
            for (const arrow of values(self.arrows)) {
                if (arrow.input === socket || arrow.output === socket) {
                    detach(arrow);
                    // arrowsToDelete.push(arrow._id);
                }
            }
            // for (const id of arrowsToDelete) {
            //     self.arrows.delete(id);
            // }
        };

        const runCode = () => {
            if (self.running) {
                if (self.breakpointCallback) {
                    self.breakpointCallback();
                    self.continueAfterBreakpoint();
                }
            } else {
                self.running = true;
                const env = getEnv<{ run: typeof run }>(self);
                env.run(self.boxes, (box, resume) =>
                    self.setBreakpoint(box, resume));
            }
        };

        const stopCode = () => {
            self.running = false;
            self.continueAfterBreakpoint();
            return getEnv(self).stop();
        };

        const addSocketToBox = (box: modelTypes['Box'], type: SocketTypeEnum, code: ICodeBlockIO) => {
            const s = models.Socket.create({ socketType: type, name: code.name, code });
            self.sockets.put(s);
            box.addSocket(s);
            return s;
        };
        const moveBoxOrSelection = (box: modelTypes['Box'], dx: number, dy: number) => {
            if (self.selection.length > 0 && self.selection.indexOf(box) > -1) {
                for (const box of self.selection) {
                    box.move(dx, dy);
                }
            } else {
                box.move(dx, dy);
            }
        };

        return {
            addBox,
            addBoxAndArrowIfDragged,
            hasArrow,
            setSelection,
            addToSelection,
            createBox,
            deleteSelection,
            deleteBox,
            addArrow,
            startDragArrow,
            moveDragArrow,
            endDragArrow,
            startDragRect,
            moveDragRect,
            endDragRect,
            deleteArrowsForSocket,
            runCode,
            stopCode,
            addSocketToBox,
            moveBoxOrSelection,
        };
    })



/*
    The store that holds our domain: boxes and arrows
*/

export const defaults: IStoreSnapshot = {
    boxes: {},
    arrows: {},
    contextMenu: {},
    codeBlocks: { ...codeBlocks.functions },
    selection: [],
    draggedArrow: null,
    draggedFromSocket: null
};

/* istanbul ignore next */
export const getStore = (data?: IStoreSnapshot) => {
    return Store.create({ ...defaults, ...data }, { run, stop });
}

export const replaceStore = (newStore: IStore, oldStore: IStore) => {
    applySnapshot(newStore, { ...getSnapshot(oldStore), codeBlocks: { ...codeBlocks.functions } });
};

type IStoreType = typeof Store.Type;
export interface IStore extends IStoreType { };
type IStoreSnapshotType = typeof Store.SnapshotType;
export interface IStoreSnapshot extends IStoreSnapshotType { };
