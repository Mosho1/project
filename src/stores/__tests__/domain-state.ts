// import { getSnapshot, applyAction } from "mobx-state-tree"
import { Store, IStore } from '../domain-state';
import { product, mock } from '../test-utils';
import { socketTypes } from '../models/socket';
import { createTestCodeBlock } from '../models/__tests__/code-block';
import { createTestBox } from '../models/__tests__/box';
import { ICodeBlock } from '../models/code-block';

let store: IStore;
beforeEach(() => {
    store = Store.create();
});

test('addBox', () => {
    const cb = createTestCodeBlock();
    const box = store.addBox('test', 0, 0, cb);
    expect(store.boxes.get(box._id)).toBe(box);
});

test('setSelection', () => {
    expect(store.setSelection(null).selection).toBeNull();
    const box = createTestBox();
    expect(store.setSelection(box).selection).toBe(box);
});

test('createBox', () => {
    const cb = createTestCodeBlock();
    const b = store.createBox('test', 0, 0, cb);
    expect(store.selection).toBe(b);
});

test('deleteBox', () => {
    const cb = createTestCodeBlock({
        execOutputs: [''],
        execInputs: [''],
        returns: 'number'
    });
    const b1 = store.addBox('test', 0, 0, cb);
    const b2 = store.addBox('test2', 50, 50, cb);
    expect(store.boxes).toHaveProperty('size', 2);
    store.addArrow(b1.outputs[0], b2.inputs[0]);
    store.addArrow(b1.execInputs[0], b2.execOutputs[0]);
    expect(store.arrows).toHaveProperty('size', 2);
    store.setSelection(b1);
    store.deleteBox(b1);
    expect(store.selection).toBeNull();
    expect(store.arrows).toHaveProperty('size', 0);
    expect(store.boxes).toHaveProperty('size', 1);
    store.deleteBox(b2);
    expect(store.boxes).toHaveProperty('size', 0);
});

test('addArrow', () => {
    const tests = product(socketTypes, socketTypes);

    for (const test of tests) {
        const [t1, t2] = test;
        const box = createTestBox();
        const s1 = box.addSocket(t1);
        const s2 = box.addSocket(t2);
        expect(store.addArrow(s1, s2) !== null).toMatchSnapshot(test.join(','));
    }
});

test('hasArrow', () => {
    const tests = product(socketTypes, socketTypes);

    for (const test of tests) {
        const [t1, t2] = test;
        const box = createTestBox();
        const s1 = box.addSocket(t1);
        const s2 = box.addSocket(t2);
        store.addArrow(s1, s2);
        expect(store.hasArrow(s1, s2)).toMatchSnapshot(test.join(','));
    }
});

test('startDragArrow', () => {
    const box = createTestBox({ x: 10, y: 10 });
    const socket = box.addSocket('input');
    store.startDragArrow(socket);
    expect(store.draggedArrow).toMatchObject({
        startX: socket.x,
        startY: socket.y,
        endX: socket.x,
        endY: socket.y,
    });
    expect(store.draggedFromSocket).toBe(socket);
});

test('moveDragArrow', () => {
    const draggedArrow = { start: () => null, end: () => null };
    const draggedFromSocket = { isInput: false };
    spyOn(draggedArrow, 'start');
    spyOn(draggedArrow, 'end');
    mock(store, { draggedArrow, draggedFromSocket });
    store.moveDragArrow(1, 2);
    draggedFromSocket.isInput = true;
    store.moveDragArrow(3, 4);
    expect(draggedArrow.end).toHaveBeenCalledWith(1, 2);
    expect(draggedArrow.start).toHaveBeenCalledWith(3, 4);
});

test('endDragArrow', () => {
    const cb = createTestCodeBlock();
    const b1 = store.addBox('test', 0, 0, cb);
    const s1 = b1.addSocket('input');

    const b2 = store.addBox('test', 0, 0, cb);
    const s2 = b2.addSocket('output');

    mock(store, { draggedArrow: {}, draggedFromSocket: s1 });
    store.endDragArrow(s2);

    expect(store.hasArrow(s1, s2)).toBe(true);
    expect(store.draggedArrow).toBeNull();
});

test('runCode', () => {
    const obj = {run: () => null};
    spyOn(obj, 'run');
    store = Store.create({}, obj);
    store.runCode();
    expect(obj.run).toHaveBeenCalledWith(store.boxes);
});

test('sortedCodeBlocks', () => {
    let codeBlocks: {[index: string]: ICodeBlock} = {};
    for (const name of ['zzz','ac','asef','ba','esv','cef','cef','ce3','dd','d']) {
        codeBlocks[name] = createTestCodeBlock({
            id: name,
            name
        });
    }
    store = Store.create({codeBlocks})
    expect(store.sortedCodeBlocks).toMatchSnapshot();
});