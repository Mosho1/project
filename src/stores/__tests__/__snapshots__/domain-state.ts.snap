// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addArrow: execInput,execInput 1`] = `false`;

exports[`addArrow: execInput,execOutput 1`] = `true`;

exports[`addArrow: execInput,input 1`] = `false`;

exports[`addArrow: execInput,output 1`] = `false`;

exports[`addArrow: execOutput,execInput 1`] = `true`;

exports[`addArrow: execOutput,execOutput 1`] = `false`;

exports[`addArrow: execOutput,input 1`] = `false`;

exports[`addArrow: execOutput,output 1`] = `false`;

exports[`addArrow: input,execInput 1`] = `false`;

exports[`addArrow: input,execOutput 1`] = `false`;

exports[`addArrow: input,input 1`] = `false`;

exports[`addArrow: input,output 1`] = `true`;

exports[`addArrow: output,execInput 1`] = `false`;

exports[`addArrow: output,execOutput 1`] = `false`;

exports[`addArrow: output,input 1`] = `true`;

exports[`addArrow: output,output 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,0,0 1`] = `true`;

exports[`areSocketsCompatible: execInput,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,0,0 1`] = `true`;

exports[`areSocketsCompatible: execOutput,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,output,0,0 1`] = `true`;

exports[`areSocketsCompatible: input,output,0,1 1`] = `true`;

exports[`areSocketsCompatible: input,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,input,0,0 1`] = `true`;

exports[`areSocketsCompatible: output,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,input,1,0 1`] = `true`;

exports[`areSocketsCompatible: output,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,output,1,1 1`] = `false`;

exports[`hasArrow: execInput,execInput 1`] = `false`;

exports[`hasArrow: execInput,execOutput 1`] = `true`;

exports[`hasArrow: execInput,input 1`] = `false`;

exports[`hasArrow: execInput,output 1`] = `false`;

exports[`hasArrow: execOutput,execInput 1`] = `false`;

exports[`hasArrow: execOutput,execOutput 1`] = `false`;

exports[`hasArrow: execOutput,input 1`] = `false`;

exports[`hasArrow: execOutput,output 1`] = `false`;

exports[`hasArrow: input,execInput 1`] = `false`;

exports[`hasArrow: input,execOutput 1`] = `false`;

exports[`hasArrow: input,input 1`] = `false`;

exports[`hasArrow: input,output 1`] = `true`;

exports[`hasArrow: output,execInput 1`] = `false`;

exports[`hasArrow: output,execOutput 1`] = `false`;

exports[`hasArrow: output,input 1`] = `false`;

exports[`hasArrow: output,output 1`] = `false`;

exports[`height: 0,0,0,0 1`] = `50`;

exports[`height: 0,0,0,1 1`] = `80`;

exports[`height: 0,0,0,2 1`] = `110`;

exports[`height: 0,0,1,0 1`] = `80`;

exports[`height: 0,0,1,1 1`] = `80`;

exports[`height: 0,0,1,2 1`] = `110`;

exports[`height: 0,0,2,0 1`] = `110`;

exports[`height: 0,0,2,1 1`] = `110`;

exports[`height: 0,0,2,2 1`] = `110`;

exports[`height: 0,1,0,0 1`] = `80`;

exports[`height: 0,1,0,1 1`] = `110`;

exports[`height: 0,1,0,2 1`] = `140`;

exports[`height: 0,1,1,0 1`] = `80`;

exports[`height: 0,1,1,1 1`] = `110`;

exports[`height: 0,1,1,2 1`] = `140`;

exports[`height: 0,1,2,0 1`] = `110`;

exports[`height: 0,1,2,1 1`] = `110`;

exports[`height: 0,1,2,2 1`] = `140`;

exports[`height: 0,2,0,0 1`] = `110`;

exports[`height: 0,2,0,1 1`] = `140`;

exports[`height: 0,2,0,2 1`] = `170`;

exports[`height: 0,2,1,0 1`] = `110`;

exports[`height: 0,2,1,1 1`] = `140`;

exports[`height: 0,2,1,2 1`] = `170`;

exports[`height: 0,2,2,0 1`] = `110`;

exports[`height: 0,2,2,1 1`] = `140`;

exports[`height: 0,2,2,2 1`] = `170`;

exports[`height: 1,0,0,0 1`] = `80`;

exports[`height: 1,0,0,1 1`] = `80`;

exports[`height: 1,0,0,2 1`] = `110`;

exports[`height: 1,0,1,0 1`] = `110`;

exports[`height: 1,0,1,1 1`] = `110`;

exports[`height: 1,0,1,2 1`] = `110`;

exports[`height: 1,0,2,0 1`] = `140`;

exports[`height: 1,0,2,1 1`] = `140`;

exports[`height: 1,0,2,2 1`] = `140`;

exports[`height: 1,1,0,0 1`] = `80`;

exports[`height: 1,1,0,1 1`] = `110`;

exports[`height: 1,1,0,2 1`] = `140`;

exports[`height: 1,1,1,0 1`] = `110`;

exports[`height: 1,1,1,1 1`] = `110`;

exports[`height: 1,1,1,2 1`] = `140`;

exports[`height: 1,1,2,0 1`] = `140`;

exports[`height: 1,1,2,1 1`] = `140`;

exports[`height: 1,1,2,2 1`] = `140`;

exports[`height: 1,2,0,0 1`] = `110`;

exports[`height: 1,2,0,1 1`] = `140`;

exports[`height: 1,2,0,2 1`] = `170`;

exports[`height: 1,2,1,0 1`] = `110`;

exports[`height: 1,2,1,1 1`] = `140`;

exports[`height: 1,2,1,2 1`] = `170`;

exports[`height: 1,2,2,0 1`] = `140`;

exports[`height: 1,2,2,1 1`] = `140`;

exports[`height: 1,2,2,2 1`] = `170`;

exports[`height: 2,0,0,0 1`] = `110`;

exports[`height: 2,0,0,1 1`] = `110`;

exports[`height: 2,0,0,2 1`] = `110`;

exports[`height: 2,0,1,0 1`] = `140`;

exports[`height: 2,0,1,1 1`] = `140`;

exports[`height: 2,0,1,2 1`] = `140`;

exports[`height: 2,0,2,0 1`] = `170`;

exports[`height: 2,0,2,1 1`] = `170`;

exports[`height: 2,0,2,2 1`] = `170`;

exports[`height: 2,1,0,0 1`] = `110`;

exports[`height: 2,1,0,1 1`] = `110`;

exports[`height: 2,1,0,2 1`] = `140`;

exports[`height: 2,1,1,0 1`] = `140`;

exports[`height: 2,1,1,1 1`] = `140`;

exports[`height: 2,1,1,2 1`] = `140`;

exports[`height: 2,1,2,0 1`] = `170`;

exports[`height: 2,1,2,1 1`] = `170`;

exports[`height: 2,1,2,2 1`] = `170`;

exports[`height: 2,2,0,0 1`] = `110`;

exports[`height: 2,2,0,1 1`] = `140`;

exports[`height: 2,2,0,2 1`] = `170`;

exports[`height: 2,2,1,0 1`] = `140`;

exports[`height: 2,2,1,1 1`] = `140`;

exports[`height: 2,2,1,2 1`] = `170`;

exports[`height: 2,2,2,0 1`] = `170`;

exports[`height: 2,2,2,1 1`] = `170`;

exports[`height: 2,2,2,2 1`] = `170`;

exports[`index tests: execInput,0,0,0,0 1`] = `-1`;

exports[`index tests: execInput,0,0,0,1 1`] = `-1`;

exports[`index tests: execInput,0,0,1,0 1`] = `-1`;

exports[`index tests: execInput,0,0,1,1 1`] = `-1`;

exports[`index tests: execInput,0,1,0,0 1`] = `-1`;

exports[`index tests: execInput,0,1,0,1 1`] = `-1`;

exports[`index tests: execInput,0,1,1,0 1`] = `-1`;

exports[`index tests: execInput,0,1,1,1 1`] = `-1`;

exports[`index tests: execInput,1,0,0,0 1`] = `-1`;

exports[`index tests: execInput,1,0,0,1 1`] = `-1`;

exports[`index tests: execInput,1,0,1,0 1`] = `-1`;

exports[`index tests: execInput,1,0,1,1 1`] = `-1`;

exports[`index tests: execInput,1,1,0,0 1`] = `-1`;

exports[`index tests: execInput,1,1,0,1 1`] = `-1`;

exports[`index tests: execInput,1,1,1,0 1`] = `-1`;

exports[`index tests: execInput,1,1,1,1 1`] = `-1`;

exports[`index tests: execOutput,0,0,0,0 1`] = `-1`;

exports[`index tests: execOutput,0,0,0,1 1`] = `-1`;

exports[`index tests: execOutput,0,0,1,0 1`] = `-1`;

exports[`index tests: execOutput,0,0,1,1 1`] = `-1`;

exports[`index tests: execOutput,0,1,0,0 1`] = `-1`;

exports[`index tests: execOutput,0,1,0,1 1`] = `-1`;

exports[`index tests: execOutput,0,1,1,0 1`] = `-1`;

exports[`index tests: execOutput,0,1,1,1 1`] = `-1`;

exports[`index tests: execOutput,1,0,0,0 1`] = `-1`;

exports[`index tests: execOutput,1,0,0,1 1`] = `-1`;

exports[`index tests: execOutput,1,0,1,0 1`] = `-1`;

exports[`index tests: execOutput,1,0,1,1 1`] = `-1`;

exports[`index tests: execOutput,1,1,0,0 1`] = `-1`;

exports[`index tests: execOutput,1,1,0,1 1`] = `-1`;

exports[`index tests: execOutput,1,1,1,0 1`] = `-1`;

exports[`index tests: execOutput,1,1,1,1 1`] = `-1`;

exports[`index tests: input,0,0,0,0 1`] = `-1`;

exports[`index tests: input,0,0,0,1 1`] = `-1`;

exports[`index tests: input,0,0,1,0 1`] = `0`;

exports[`index tests: input,0,0,1,1 1`] = `0`;

exports[`index tests: input,0,1,0,0 1`] = `-1`;

exports[`index tests: input,0,1,0,1 1`] = `-1`;

exports[`index tests: input,0,1,1,0 1`] = `0`;

exports[`index tests: input,0,1,1,1 1`] = `0`;

exports[`index tests: input,1,0,0,0 1`] = `-1`;

exports[`index tests: input,1,0,0,1 1`] = `-1`;

exports[`index tests: input,1,0,1,0 1`] = `0`;

exports[`index tests: input,1,0,1,1 1`] = `0`;

exports[`index tests: input,1,1,0,0 1`] = `-1`;

exports[`index tests: input,1,1,0,1 1`] = `-1`;

exports[`index tests: input,1,1,1,0 1`] = `0`;

exports[`index tests: input,1,1,1,1 1`] = `0`;

exports[`index tests: output,0,0,0,0 1`] = `-1`;

exports[`index tests: output,0,0,0,1 1`] = `0`;

exports[`index tests: output,0,0,1,0 1`] = `-1`;

exports[`index tests: output,0,0,1,1 1`] = `0`;

exports[`index tests: output,0,1,0,0 1`] = `-1`;

exports[`index tests: output,0,1,0,1 1`] = `0`;

exports[`index tests: output,0,1,1,0 1`] = `-1`;

exports[`index tests: output,0,1,1,1 1`] = `0`;

exports[`index tests: output,1,0,0,0 1`] = `-1`;

exports[`index tests: output,1,0,0,1 1`] = `0`;

exports[`index tests: output,1,0,1,0 1`] = `-1`;

exports[`index tests: output,1,0,1,1 1`] = `0`;

exports[`index tests: output,1,1,0,0 1`] = `-1`;

exports[`index tests: output,1,1,0,1 1`] = `0`;

exports[`index tests: output,1,1,1,0 1`] = `-1`;

exports[`index tests: output,1,1,1,1 1`] = `0`;

exports[`move 1`] = `
Object {
  "_id": "test",
  "code": "cbId",
  "mstPouchType": "Box",
  "name": "hal",
  "sockets": Array [],
  "values": Array [],
  "x": 1,
  "y": 2,
}
`;

exports[`move 2`] = `
Object {
  "_id": "test",
  "code": "cbId",
  "mstPouchType": "Box",
  "name": "hal",
  "sockets": Array [],
  "values": Array [],
  "x": 2,
  "y": 4,
}
`;

exports[`x/y execInput 1`] = `
Object {
  "x": 18,
  "y": 50,
}
`;

exports[`x/y execInput 2`] = `
Object {
  "x": -82,
  "y": 150,
}
`;

exports[`x/y execInput 3`] = `
Object {
  "x": -82,
  "y": 210,
}
`;

exports[`x/y execInput 4`] = `
Object {
  "x": 18,
  "y": 240,
}
`;

exports[`x/y execInput 5`] = `
Object {
  "x": -482,
  "y": 200,
}
`;

exports[`x/y execOutput 1`] = `
Object {
  "x": -18,
  "y": 50,
}
`;

exports[`x/y execOutput 2`] = `
Object {
  "x": -68,
  "y": 150,
}
`;

exports[`x/y execOutput 3`] = `
Object {
  "x": -68,
  "y": 210,
}
`;

exports[`x/y execOutput 4`] = `
Object {
  "x": 2,
  "y": 240,
}
`;

exports[`x/y execOutput 5`] = `
Object {
  "x": -418,
  "y": 200,
}
`;

exports[`x/y input 1`] = `
Object {
  "x": 18,
  "y": 50,
}
`;

exports[`x/y input 2`] = `
Object {
  "x": -82,
  "y": 150,
}
`;

exports[`x/y input 3`] = `
Object {
  "x": -82,
  "y": 210,
}
`;

exports[`x/y input 4`] = `
Object {
  "x": 18,
  "y": 240,
}
`;

exports[`x/y input 5`] = `
Object {
  "x": -482,
  "y": 200,
}
`;

exports[`x/y output 1`] = `
Object {
  "x": -18,
  "y": 50,
}
`;

exports[`x/y output 2`] = `
Object {
  "x": -68,
  "y": 150,
}
`;

exports[`x/y output 3`] = `
Object {
  "x": -68,
  "y": 210,
}
`;

exports[`x/y output 4`] = `
Object {
  "x": 2,
  "y": 240,
}
`;

exports[`x/y output 5`] = `
Object {
  "x": -418,
  "y": 200,
}
`;
